Add Govee local light integration

Galorhallen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR introduces a new integrations: Govee lights local.

This integration allows Home Assistant to control all Govee lights that supports local api control. This feature has been added in the newer Govee devices and allows the devices to be controlled directly, without cloud, on the local network via UDP messages.

The integration has a discovery mechanism, based on the Govee API, that automatically adds new device to Home Assistant when they are connected to the local network and the local control is enabled.
The integration also update the status of the entity if the light configuration is changed externally (i.e. via Govee App)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

The integrations uses [this library](https://pypi.org/project/govee-local-api/) ([source](https://github.com/Galorhallen/govee-local-api)) specifically developed for this integration.
The library is based on the Govee Local API documentation (https://app-h5.govee.com/user-manual/wlan-guide)

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30564

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Galorhallen committed:
Added govee_local_api

Galorhallen committed:
Code cleanup

Galorhallen committed:
Fix discovery

Galorhallen committed:
Add missing supported device

Galorhallen committed:
Fix autodiscovery

Galorhallen committed:
Add missing quality scale in manifest.json

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,213 @@
+    """Govee light setup."""
+
+    entry_id = config_entry.entry_id
+    coordinator: GoveeLocalApiCoordinator = hass.data.setdefault(DOMAIN, {}).get(
+        entry_id, None
+    )
```
Can we move this to `__init__.py`?

Galorhallen replied:
Do you mean to move the entire code of async_setup_entry in the one in the __init__.py?

joostlek replied:
We want to create the coordinator in the init file, not in a platform file

joostlek commented on homeassistant/components/govee_light_api/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+_LOGGING = logging.getLogger(__name__)
+
+
+class GoveeConfigFlow(ConfigFlow, domain=DOMAIN):
```
Since we're only doing UDP discovery, isn't a config flow like Gree better? Are there reasons why someone has a different port for UDP discovery?

Galorhallen replied:
Yeah, probably chainging the ports (or even the multicast address) is not necessary, since they cannot be changed on the Govee side.
I'll check the Gree flow and update this one

Galorhallen replied:
Regarding this, I'm almost done, but I didn't find a way to keep the OptionsFlow.
Is there a way to keep it? It's not really needed but we will lose the possibility to change the discovery interval.

Also, looking at the Gree integration, I had a couple of ideas:
* Would be better to fire an event when a device is discovered?
* Currently the discovery it is done using the `govee_local_api` library (internally it uses a `call_later` on the same Home Assistant event loop) but, since this can be disabled, could be better if the update function is triggered using something like `async_track_time_interval`?

joostlek replied:
Why would one change the discovery interval?

What do you mean with the config flow can be disabled?

joostlek replied:
Also, Optionsflow should be in a follow up PR. If with the gree config flow no options flow is possible and we still want one, I'd rather fix something in core so we are able to add an options flow

Galorhallen replied:
> Why would one change the discovery interval?

That is probably not really needed, unless maybe for debug purposes.

> What do you mean with the config flow can be disabled?

I didn't mean the config flow, I mixed up things there.
What I meant is that the library internally uses `call_later` to poll the Govee devices and update their status, but this can be disabled and the devices can be polled "manually".
I was wondering if it should be better to let Home Assistant call the poll function.

joostlek commented on homeassistant/components/govee_light_api/coordinator.py:
```diff
@@ -0,0 +1,122 @@
+            update_interval=scan_interval,
+        )
+
+        self._config_entry = config_entry
```
after the `super().__init__()`, `self.config_entry` is set to the config entry. Please retype `config_entry` to `ConfigEntry` tho. Checkout the youtube coordinator on how to do that

joostlek commented on homeassistant/components/govee_light_api/coordinator.py:
```diff
@@ -0,0 +1,122 @@
+
+    async def turn_on(self, device: GoveeDevice) -> None:
+        """Turn on the light."""
+        assert self._controller == device.controller
```
why do we do this?

Galorhallen replied:
This is an excellent question 😅 I remember writing this in an early draft of the code, but it makes little sense now

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,213 @@
+    ) -> None:
+        """Govee Light constructor."""
+
+        super(CoordinatorEntity, self).__init__(coordinator)
```
```suggestion
        super().__init__(coordinator)
```

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,213 @@
+
+        super(CoordinatorEntity, self).__init__(coordinator)
+
+        self._coordinator = coordinator
```
`CoordinatorEntity` will set the `coordinator` to `self.coordinator`

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,213 @@
+        )
+
+        self._attr_unique_id: str = device.fingerprint
+        self.entry_id = f"{device.sku}_{device.fingerprint}"
```
```suggestion
```

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,213 @@
+        """Name of the entity."""
+        return self._device.sku
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return self._attr_unique_id
```
The default implementation is already returning self._attr_unique_id

joostlek commented on homeassistant/components/govee_light_api/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "config_flow": true,
+  "dependencies": ["network"],
+  "documentation": "https://www.home-assistant.io/integrations/govee_light_api",
+  "homekit": {},
+  "iot_class": "local_push",
+  "quality_scale": "silver",
+  "requirements": ["govee-local-api==1.3.2"],
+  "ssdp": [],
+  "zeroconf": []
```
Please remove empty fields. Also remove the quality scale. I would prefer to have that done in a separate PR

joostlek reviewed (changes requested):
This is nice, I have strings lights of them and use a HACS integration for this

klaasnicolaas commented:
Why not just call it `govee_light` as name/domain?

Galorhallen committed:
QA

Galorhallen commented:
> Why not just call it `govee_light` as name/domain?

I named this way because older Govee lights uses BLE to control them, since they don't have the local API interface.
I'm working on another integration for them, but it is kind of hackish since the protocol is not documented and I'm not sure if it will land somewhere.

joostlek commented:
Is there a big difference between those and govee_ble?

Galorhallen commented:
> Is there a big difference between those and govee_ble?

If I remember correctly my first attempt was to extend govee_ble, but I ended up starting from scratch. I'm not sure what was the problem, it was a while ago.
Probably I'll give it another try after this one.

Galorhallen committed:
QA: Moved coordinator creation to __init__.py

joostlek commented:
Please only request review by pressing "ready for review"

joostlek commented on .devcontainer/devcontainer.json:
```diff
@@ -6,7 +6,7 @@
   "postStartCommand": "script/bootstrap",
   "containerEnv": { "DEVCONTAINER": "1" },
   "appPort": ["8123:8123"],
-  "runArgs": ["-e", "GIT_EDITOR=code --wait"],
```
Please revert this

joostlek commented on homeassistant/components/govee_light_api/__init__.py:
```diff
@@ -0,0 +1,56 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Govee Local API from a config entry."""
+
+    entry_id = entry.entry_id
+    coordinator: GoveeLocalApiCoordinator = hass.data.setdefault(DOMAIN, {}).get(
+        entry_id, None
+    )
```
Coordinator is not setup at this point, like never, so we can remove it here.

joostlek commented on homeassistant/components/govee_light_api/__init__.py:
```diff
@@ -0,0 +1,56 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
This unload is lacking unloading platforms

joostlek commented on homeassistant/components/govee_light_api/__init__.py:
```diff
@@ -0,0 +1,56 @@
+        entry_id
+    )
+    if coordinator:
+        coordinator.clenaup()
```
You can add `entry.async_on_unload(coordinator.cleanup)` in the setup, this removes the need to do it here as well

joostlek commented on homeassistant/components/govee_light_api/__init__.py:
```diff
@@ -0,0 +1,56 @@
+    if coordinator:
+        coordinator.clenaup()
+        del hass.data[DOMAIN][entry_id]
+        del hass.data[DOMAIN]
```
Please check other unload entries, as this is wrong

joostlek commented on homeassistant/components/govee_light_api/coordinator.py:
```diff
@@ -0,0 +1,117 @@
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+        scan_interval: timedelta,
+        logger: logging.Logger,
```
These 3 parameters are unneeded imo

joostlek commented on homeassistant/components/govee_light_api/coordinator.py:
```diff
@@ -0,0 +1,117 @@
+            update_interval=scan_interval,
+        )
+
+        config = config_entry.data["config"]
```
This isn't doing anything useful?

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,182 @@
+
+    coordinator: GoveeLocalApiCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    await coordinator.async_config_entry_first_refresh()
```
This should go in the init file

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,182 @@
+
+    await coordinator.async_config_entry_first_refresh()
+
+    try:
+        async with asyncio.timeout(delay=5):
+            while not coordinator.devices:
+                await asyncio.sleep(delay=1)
+    except asyncio.TimeoutError:
+        pass
```
This too

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,182 @@
+            device.sku, set()
+        )
+
+        self._attr_unique_id: str = device.fingerprint
```
```suggestion
        self._attr_unique_id = device.fingerprint
```

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,182 @@
+        if len(color_modes) == 0:
+            color_modes.add(ColorMode.ONOFF)
+
+        self._attr_supported_color_modes: set[ColorMode] | set[str] = color_modes
```
```suggestion
        self._attr_supported_color_modes = color_modes
```

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,182 @@
+
+        self._attr_supported_color_modes: set[ColorMode] | set[str] = color_modes
+
+    @property
+    def name(self) -> str:
+        """Name of the entity."""
+        return self._device.sku
```
device_info can be set in the constructor with `_attr_device_info`. Then you can set `_attr_name = None` and the entity will follow the device name

Galorhallen committed:
QA

joostlek commented:
Please don't force push as it makes it harder to see what changes in between changes

joostlek commented on homeassistant/components/govee_light_api/coordinator.py:
```diff
@@ -0,0 +1,93 @@
+        """Set discovery callback for automatic Govee light discovery."""
+        self._controller.set_device_discovered_callback(callback)
+
+    def clenaup(self) -> None:
+        """Stop and clenaup the cooridinator."""
+        self._controller.clenaup()
```
It's cleanup. Please fix this in the code or it'll probably stay here forever

Galorhallen replied:
I was able to get it wrong also in the library 😓 I'll also fix there

Galorhallen committed:
Fix typo and update test

joostlek commented on homeassistant/components/govee_light_api/__init__.py:
```diff
@@ -0,0 +1,51 @@
+    """Set up Govee Local API from a config entry."""
+
+    coordinator: GoveeLocalApiCoordinator = GoveeLocalApiCoordinator(
+        hass=hass, scan_interval=SCAN_INTERVAL, logger=_LOGGER
```
Again, these values are static, no need to pass them to the coordinator. If you disagree with something please respond instead of ignoring it.

Galorhallen replied:
Apologies, I didn't mean to ignore the comment. I guess I forgot to remove them all form the constructor, but I agree, there is no need to pass them to the constructor.

joostlek commented on homeassistant/components/govee_light_api/__init__.py:
```diff
@@ -0,0 +1,51 @@
+            while not coordinator.devices:
+                await asyncio.sleep(delay=1)
+    except asyncio.TimeoutError:
+        _LOGGER.debug("No devices found")
```
Should we raise ConfigEntryNotReady so it will try again?

joostlek commented on homeassistant/components/govee_light_api/__init__.py:
```diff
@@ -0,0 +1,51 @@
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+
+    if await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id, None)
+        return True
+    return False
```
You can use a walrus for this

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,173 @@
+        if is_new:
+            entity = GoveeLight(coordinator, device)
+            async_add_entities([entity])
+            async_dispatcher_send(hass, DISPATCH_GOVEE_LIGHT_DISCOVERED, entity)
```
Why do we dispatch this?

Galorhallen replied:
I took it from the other integration, but if it's not required or needed I'll remove it

Galorhallen replied:
Oh, I take another look at the code and I totally misunderstood the use of `async_dispatcher_send`. I'll remove it

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,173 @@
+            color_modes.add(ColorMode.ONOFF)
+
+        self._attr_supported_color_modes = color_modes
+        self._attr_name = None
```
Can be set outside of the constructor

Galorhallen committed:
QA

Galorhallen committed:
Removed unecessary code

autinerd commented on homeassistant/components/govee_light_api/const.py:
```diff
@@ -0,0 +1,13 @@
+
+CONF_MULTICAST_ADDRESS_DEFAULT = "239.255.255.250"
+CONF_TARGET_PORT_DEFAULT = 4001
+CONF_LISENING_PORT_DEFAULT = 4002
```
Typo in const identifier
```suggestion
CONF_LISTENING_PORT_DEFAULT = 4002
```

autinerd commented on homeassistant/components/govee_light_api/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+
+from .const import (
+    CONF_DISCOVERY_INTERVAL_DEFAULT,
+    CONF_LISENING_PORT_DEFAULT,
```
```suggestion
    CONF_LISTENING_PORT_DEFAULT,
```

autinerd commented on homeassistant/components/govee_light_api/coordinator.py:
```diff
@@ -0,0 +1,90 @@
+            logger=_LOGGER,
+            broadcast_address=CONF_MULTICAST_ADDRESS_DEFAULT,
+            broadcast_port=CONF_TARGET_PORT_DEFAULT,
+            listening_port=CONF_LISENING_PORT_DEFAULT,
```
```suggestion
            listening_port=CONF_LISTENING_PORT_DEFAULT,
```

autinerd commented on homeassistant/components/govee_light_api/strings.json:
```diff
@@ -0,0 +1,19 @@
+        "data": {
+          "multicast_address": "Multicast discovery address",
+          "target_port": "API devices port",
+          "listening_port": "API listing port",
```
```suggestion
          "listening_port": "API listening port",
```

Galorhallen committed:
Fix typo

Galorhallen committed:
Fix typo

joostlek commented on homeassistant/components/govee_light_api/strings.json:
```diff
@@ -0,0 +1,19 @@
+{
```
Please check the gree strings

Galorhallen commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,171 @@
+        return self._device.on
+
+    @property
+    def brightness(self) -> int | None:
+        """Return the brightness of this light between 0..255."""
+        return int((self._device.brightness / 100.0) * 255.0)
```
I checked the API and it looks like it always returns the last brightness value when the device is off

Galorhallen committed:
QA, typing and strings

Galorhallen committed:
Removed unsed logger in __init__.py

Galorhallen commented:
While I was double-checking the APIs, I began to doubt whether the "capabilities" defined in the `capability.py` are actually necessary.
The file only lists the **tested** devices, but the APIs should work with all devices listed in their documentation.
I think it might make sense to remove the capabilities, assuming all devices have them, thus support all devices listed in the Govee API's docs (currently if a device is not in the capabilities map, the integration only allows to turn the device on and off).
After all, in the worst case, if a device doesn't expose a certain functionality nothing will happen.
What do you guys think?

joostlek commented:
I mean it is a good idea to only show features which are supported. If a device doesn't support a feature we should disable it. (that's what its for, if it wasnt needed we wouldnt have these functionality at all).

I didn't like the capabilities file from the beginning as its only used for lights, so it could be combined. But I wanted to give it a try at my string lights before complaining about it :P

autinerd commented:
My question would be, is there a way to get the capabilities without hardcoding the models?

autinerd commented on homeassistant/components/govee_light_api/capability.py:
```diff
@@ -0,0 +1,38 @@
+from typing import Any
+
+
+class GoveeLightCapabilities(Enum):
+    """Govee Lights capabilities."""
+
+    COLOR_RGB = auto()
+    COLOR_KELVIN_TEMPERATURE = auto()
+    BRIGHTNESS = auto()
```
Is this extra enum needed or can we not just use the existing `ColorMode`?

Galorhallen commented:
> I mean it is a good idea to only show features which are supported. If a device doesn't support a feature we should disable it. (that's what its for, if it wasnt needed we wouldnt have these functionality at all).

I agree. Maybe it could be done the other way around: the integration could assume that all devices have all the capabilities and have a `not_supported_capabilities` map to be filled in future releases if someone discovers a device lacking certain functionalities.

> My question would be, is there a way to get the capabilities without hardcoding the models?

Sadly no, the APIs expose almost no information on the devices.

joostlek commented:
I mean, we know the list of supported devices (that use the UDP protocol), so we can do some googling and do some assumptions and we can get issues if they are incorrect

Galorhallen committed:
QA, using ColorMode for lights capabilities

joostlek commented:
by the way, could we store the capabilities in the lib? And return our own enum which we transform to an HA enum in the integration?

Galorhallen commented:
Yeah, storing it in the library could be better.
I'll check on the internet for the supported devices capabilities and update the library.

autinerd commented:
> Sadly no, the APIs expose almost no information on the devices.
> 
> 

So the `devStatus` always returns color and colorTemp even on devices which don't support these?

Galorhallen commented:
It's not clearly specified in the docs, but it looks like the structure of the JSON seems to be pretty static.
Doing some tests, it looks like that the behavior of the APIs is:
* They always return the last configured value for RGB, temperature and brightness, even if the device is turned off.
* If the color temperature is configured, the RGB fields go to zero, and vice versa.

autinerd commented:
> It's not clearly specified in the docs, but it looks like the structure of the JSON seems to be pretty static and there is no reference on how the JSON should look like if the device doesn't support a specific feature.
> 

Yeah, I understand, the only method of getting that information is testing on a real device and for that we need somebody who has such a device with only limited capabilities.

We can begin with a lookup table of course (but in the underlying lib, not directly in the integration) and put a entry into the log when we encounter a device which is not in the lookup table (like `cast` does). And if it turns out that the capabilities can be seen from the API, then the lib can be modified accordingly.

Galorhallen committed:
Bump govee_local_api to 1.4.0

Moved capabilities to library.

Galorhallen committed:
Update requirements

Galorhallen committed:
Merge branch 'dev' into govee-local-api

joostlek commented:
Are all devices in yet? I just tried it on my dev env on my own H70C1, and I can only turn it on and off

Galorhallen commented:
Yes, they are. But I don't see you model in the list on the APIs docs.
At this point, I'm not sure if the list is kept really up to date and all newer/updated devices are getting the local API.

Compiling the list i found only a couple of devices that supports only brightness.
Maybe the default should be to have all features and maintain a list of devices that have limited capabilities. What do you think?

joostlek commented:
I'd say the other way around. Since we retrieve capabilities from the lib, can we add a logger warning in the lib asking for people to open an issue at the library. We can allow every device to do everything, but people are less eager to mention broken features than features they actually want. I think this is the quickest way to get to feature complete.

Galorhallen commented:
Ok, so I'll add the log line to the lib and update the dependencies.
I guess I can also add your device model to the list.

Galorhallen committed:
Update library to 1.4.1 with unsupported dvice warning

Galorhallen committed:
Merge branch 'govee-local-api' of github.com:Galorhallen/core into govee-local-api

Galorhallen commented:
Ok, I've updated the library with a warning message for unsupported devices that points to the library issue page and add I've added the `H70C1` to the list of devices.

Galorhallen committed:
Fix tests after library update

Galorhallen committed:
Merge branch 'dev' into govee-local-api

joostlek commented on homeassistant/components/govee_light_api/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+
+    devices_count = len(controller.devices)
+    controller.cleanup()
+    controller = None
```
```suggestion
```

joostlek commented on homeassistant/components/govee_light_api/light.py:
```diff
@@ -0,0 +1,161 @@
+                color_modes.add(ColorMode.RGB)
+            if GoveeLightCapability.COLOR_KELVIN_TEMPERATURE in capabilities:
+                color_modes.add(ColorMode.COLOR_TEMP)
+                if ColorMode.COLOR_TEMP in color_modes:
```
This if statement doesn't do anything? This expression is always made True the line before

joostlek commented on tests/components/govee_light_api/test_light.py:
```diff
@@ -0,0 +1,315 @@
+        # Remove
+        assert await hass.config_entries.async_remove(entry.entry_id)
+        await hass.async_block_till_done()
+        assert len(hass.states.async_all()) == 0
```
I'd rather test if the "light.something" is not in hass.states

joostlek commented on tests/components/govee_light_api/test_light.py:
```diff
@@ -0,0 +1,315 @@
+
+        assert await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+        assert len(hass.states.async_all()) == 1
```
```suggestion
```

joostlek reviewed (changes requested):
I'll give this another test run

Galorhallen committed:
QA

Galorhallen committed:
Add test for retry config

autinerd commented:
For me it looks good now (of course I can't test because I don't have these devices)

joostlek reviewed (approved):
I think this PR is good. It works on my lights now as well. The only thing I don't really like is the domain name and integration name, but I have no better alternative. I will ask around for any improvements for that.

joostlek commented on homeassistant/components/govee_light_api/manifest.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "domain": "govee_light_api",
+  "name": "Govee Lights - Local API",
```
```suggestion
  "domain": "govee_light_local",
  "name": "Govee lights local",
```
Discussed with some people and we came up with these names. If we can change this in this PR, that'd be great

Galorhallen committed:
Update integration name and domain

Galorhallen commented:
I've renamed the integration and domain. I've updated docs and brand PR as well.

Galorhallen committed:
Merge branch 'dev' into govee-local-api

joostlek commented on homeassistant/components/govee_light_local/light.py:
```diff
@@ -0,0 +1,160 @@
+
+        super().__init__(coordinator)
+        self._device = device
+        self._device.set_update_callback(self._update_callback)
```
```suggestion
        device.set_update_callback(self._update_callback)
```

joostlek committed:
Update homeassistant/components/govee_light_local/light.py
